src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import ring_theory.integral_domain
+
+/-!
+# Multiplicative characters of finite rings and fields
+
+Let `R` be a finite commutative ring.
+A *multiplicative character* of `R` with values in `R'` is a morphism of
+monoids with zero from the multiplicative monoid of `R` into that of `R'`.
+
+We use the namespace `mul_char` for the definitions and results.
+
+## Tags
+
+multiplicative character
+-/
+
+section multiplicative
+
+namespace mul_char
+
+/-!
+### Definitions related to and results on multiplicative characters
+-/
+
+universes u v w
+
+-- The domain of our multiplicative characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial multiplicative character. It takes the value `0` on non-units and
+the value `1` on units. -/
+@[simps]
+def trivial [nontrivial R] [decidable_pred (λ x : R, is_unit x)] : R →*₀ R' :=
+{ to_fun := λ x, if is_unit x then 1 else 0,
+  map_zero' := by simp only [not_is_unit_zero, if_false],
+  map_one' := by simp only [is_unit_one, if_true],
+  map_mul' := by { intros x y,
+                   simp only [is_unit.mul_iff, boole_mul],
+                   split_ifs; tauto, } }
+attribute [protected] trivial
+
+/-- A multiplicative character takes unit values on units -/
+lemma is_unit_of_is_unit (χ : R →*₀ R') {a : R} (ha : is_unit a) : is_unit (χ a) :=
+begin
+  obtain ⟨u, rfl⟩ := ha,
+  exact is_unit_of_mul_eq_one _ _
+          (by rw [← map_mul, units.mul_inv, map_one] : χ u * χ (u⁻¹ : Rˣ) = 1),
+end
+
+/-- A multiplicative character is *nontrivial* if it takes a value `≠ 1` on a unit -/
+def is_nontrivial (χ : R →*₀ R') : Prop := ∃ (a : R), is_unit a ∧ χ a ≠ 1
```

Comment:
I would like to be able to write `χ.is_nontrivial` (or even `χ.nontrivial`). Does anyone see a simple way of getting this? Note that `χ` is a `monoid_with_zero_hom`, but we are in the `mul_char` namespace (and I think this is reasonable, even if at the end of the day a `mul_char` is just a `monoid_with_zero_hom` we don't want to use this not-so-friendly name. Will `notation` or `abbreviation` work?